build.assets/tooling/cmd/update-plist-version/main.go

```diff
@@ -0,0 +1,140 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Command update-plist-version updates the version fields of a bundle plist.
+//
+// The version fields updated in an Info.plist file are the CFBundleVersion and
+// CFBundleShortVersionString fields. If the version is not valid as per the
+// Apple specification then the fields will be set to 1.0. A valid version is 3
+// positive integers separated by dots. A semver with a pre-release tag is not
+// valid.
+//
+// This is intended to be used on the tsh.app Info.plist files. Standard
+// releases have the required version number. Pre-releases do not.
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"regexp"
+
+	"howett.net/plist"
+)
+
+func main() {
+	if len(os.Args) < 3 {
+		fmt.Fprintf(os.Stderr, "Usage: %s <version> <plist-file>...\n", filepath.Base(os.Args[0]))
+		os.Exit(1)
+	}
+
+	// A version can only be three positive integers separated by periods.
+	versionRE := regexp.MustCompile(`^[0-9]+\.[0-9]+\.[0-9]+$`)
+	version := os.Args[1]
+	if !versionRE.MatchString(version) {
+		version = "1.0"
+	}
+
+	for _, filename := range os.Args[2:] {
+		err := replaceFile(filename, func(in, out *os.File) error {
+			return updateVersion(version, in, out)
+		})
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "could not update version of %s: %v\n", filename, err)
+			os.Exit(1)
+		}
+	}
+}
+
+func updateVersion(version string, in, out *os.File) error {
+	dict := map[string]any{}
+	d := plist.NewDecoder(in)
+	if err := d.Decode(&dict); err != nil {
+		return err
+	}
+
+	if _, ok := dict["CFBundleVersion"]; !ok {
+		return errors.New("CFBundleVersion not in plist file")
+	}
+	if _, ok := dict["CFBundleShortVersionString"]; !ok {
+		return errors.New("CFBundleShortVersionString not in plist file")
+	}
+	dict["CFBundleVersion"] = version
+	dict["CFBundleShortVersionString"] = version
+
+	e := plist.NewEncoder(out)
+	e.Indent("\t")
+	if err := e.Encode(dict); err != nil {
+		return err
+	}
+	// the plist encoder does not write a newline at the end of the last line.
+	_, err := out.Write([]byte{'\n'})
+	return err
+}
+
+// Replace calls fn to replace the contents of filename. fn is passed
+// the input file and the output file and should return when it has
+// finished writing to out or an error occurs. The file will be
+// replaced atomically on success, or will be left untouched on
+// error. File permission of filename is retained, ownership and
+// attributes are not.
+func replaceFile(filename string, fn func(in, out *os.File) error) (rerr error) {
+	inf, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+	// defer inf.Close() and only return the error if not returning another error.
+	defer func() {
+		if err := inf.Close(); err != nil && rerr == nil {
+			rerr = err
+		}
+	}()
+
+	outdir := filepath.Dir(filename)
+	outname := filepath.Base(filename)
+	outf, err := os.CreateTemp(outdir, outname)
```

Comment:
Create the tempfile on the default temp dir, so if anything goes wrong we don't have leftovers in the source tree?

```suggestion
	outf, err := os.CreateTemp("", outname)
```

Reply:
If the default temp dir is on a different filesystem to the target file, we will not be able to atomically rename it. It needs to be in the same directory (same filesystem, but the only way to know without actually checking filesystems is to use the same directory)

Reply:
done